WordPress/openverse-frontend
# Frontend tokens

## Fixes


Fixes #1558 by @sarayourfriend 

## Description


This PR adds a mechanism to send authenticated requests when an API key has been provided (and unauthenticated ones) when not.

This means that API calls during SSR will be authenticated, and not throttled, but API calls from the client (subsequent calls) will be throttled as an authenticated user because they won't have any API key set.

## Testing Instructions

0. Check out the repo.
1. Follow instructions in `TESTING_GUIDELINES.md` for setting up your local environment with API authentication tokens after registering an app on your local dev server or on staging.
    * Be sure to use the version of the document from this branch as I've introduced new documentation for how to do this!
2. Visit the application as normal and confirm in the logs you can see the API token flow. Especially confirm that the _first_ request does wait for the token to come back before continuing.

Furthermore, test using the docker image:

```
docker build . --build-arg RELEASE=v0-test --build-arg API_URL=https://api-dev.openverse.engineering/ -t openverse-frontend:latest
docker run --env-file .env -p 8443:8443/tcp openverse-frontend:latest
```

> **Note:** Earlier this testing instructions read to use a locally running API with the docker container. This doesn't work due to CORS issues and having different URLs for the client and server to hit the API (due to the different network contexts they're running in). To test docker, please use a live API instance. To test against a local API, run the Nuxt app outside of docker using `pnpm prod` and an `.env` file set up to point to your local API.

Make sure to visit the docker-running service at `localhost:8443` and _not_ the IP address that nuxt spits out. That IP address is internal to the docker network. For some reason it still lets some requests through but most of them will fail (including static assets). If you stick to `localhost:8443` it works fine though.

When making requests to the docker image, you will see more than one API token being generated now in the logs, depending on the number of CPUs docker has access to. That is because pm2 runs a worker per CPU it finds. If you're using docker desktop and it is not configured with multiple CPUs then you will not see this behavior. Please enable at least 2 CPUs so that you can test this!

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
